torch/nn/modules/activation.py

```diff
@@ -1037,6 +1058,40 @@ def forward(self, query: Tensor, key: Tensor, value: Tensor, key_padding_mask: O
             `batch_first` argument is ignored for unbatched inputs.
         """
         is_batched = query.dim() == 3
+        if (torch.is_inference_mode_enabled() and is_batched and not self.training
+            and self.batch_first and self.bias_k is None and
+            self.bias_v is None and self.dropout == 0 and
+            not self.add_zero_attn and self._qkv_same_embed_dim and
+            ((key_padding_mask is None and attn_mask is None)
+             if query.is_nested
+             else (key_padding_mask is None or attn_mask is None)) and
+            query is key and key is value):
+            tensor_args = (
+                query,
+                key,
+                value,
+                self.in_proj_weight,
+                self.in_proj_bias,
+                self.out_proj.weight,
+                self.out_proj.bias,
+            )
+            if (not torch.overrides.has_torch_function(tensor_args) and
+                # We have to use a list comprehension here because
+                # Torchscript doesn't support generator expressions.
+                all([x.is_cuda or 'cpu' in str(x.device) for x in tensor_args])):
+                return torch.ops.nativetransformers._native_multi_head_attention(
+                    query,
+                    key,
+                    value,
+                    self.embed_dim,
+                    self.num_heads,
+                    self.in_proj_weight,
+                    self.in_proj_bias,
+                    self.out_proj.weight,
+                    self.out_proj.bias,
+                    key_padding_mask if key_padding_mask is not None else attn_mask,
+                    need_weights,
+                    average_attn_weights)
```

Comment:
Throw on the slow path if a NestedTensor is passed, because it doesn't support it.

Reply:
Bonus points for printing what would need to change for the fast path to hit. Then the user can construct a NestedTensor and learn what was passed vs. expected to hit the fast path / get NestedTensor support (since in this case they're the same).

Reply:
> Bonus points for printing what would need to change for the fast path to hit

I've given it some thought, but I can't see how to do that without either duplicating the list of conditions or adding more overhead.

Reply:
figured it out